integrationTests/state/genesisState_test.go

```diff
@@ -0,0 +1,304 @@
+package state
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"os"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/logger"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/data/state/addressConverters"
+	"github.com/ElrondNetwork/elrond-go/data/trie"
+	"github.com/ElrondNetwork/elrond-go/hashing/sha256"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/stretchr/testify/assert"
+)
+
+var log = logger.DefaultLogger()
+var genesisFile = "genesisBad.json"
+
+type InitialBalance struct {
+	PubKey  string `json:"pubkey"`
+	Balance string `json:"balance"`
+}
+
+type Genesis struct {
+	InitialBalances []*InitialBalance `json:"initialBalances"`
+}
+
+type testPair struct {
+	key []byte
+	val []byte
+}
+
+func TestCreationOfTheGenesisState(t *testing.T) {
+	if testing.Short() {
+		t.Skip("this is not a short test")
+	}
+
+	genesisBalances := &Genesis{}
+	err := core.LoadJsonFile(genesisBalances, genesisFile, log)
+
+	assert.Nil(t, err)
+
+	fmt.Printf("Loaded %d entries...\n", len(genesisBalances.InitialBalances))
+
+	referenceRootHash, adbReference := getRootHashByRunningInitialBalances(genesisBalances.InitialBalances)
+	fmt.Printf("Root hash: %s\n", base64.StdEncoding.EncodeToString(referenceRootHash))
+
+	_, _ = adbReference.RootHash()
+
+	noOfTests := 1000
+	for i := 0; i < noOfTests; i++ {
+		rootHash, adb := getRootHashByRunningInitialBalances(genesisBalances.InitialBalances)
+		if !bytes.Equal(rootHash, referenceRootHash) {
+			_, _ = adb.RootHash()
+			fmt.Printf("**** Wrong root hash on iteration %d: %s\n", i, base64.StdEncoding.EncodeToString(rootHash))
+			assert.Fail(t, "wrong root hash encountered")
+			return
+		}
+	}
+}
+
+func TestExtensionNodeToBranchVariant1(t *testing.T) {
+	marsh := &marshal.JsonMarshalizer{}
+	hasher := sha256.Sha256{}
+
+	tr1, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+	tr2, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+
+	key1 := "e353dd9e3da522eb366e751346195a10"
+	key2 := "a5dfc2ec3b0607e820ad375c5074c510"
+	key3 := "eb6d6e15652c0c4d1f73490e12c8b310"
+	val := "value"
+
+	_ = tr1.Update([]byte(key1), []byte(val))
+	_ = tr1.Update([]byte(key2), []byte(val))
+	_ = tr1.Update([]byte(key3), []byte(val))
+
+	fmt.Println()
+	tr1.Print(os.Stdout)
+	fmt.Println()
+
+	hash1, _ := tr1.Root()
+	fmt.Printf("root hash1: %s\n", base64.StdEncoding.EncodeToString(hash1))
+
+	_ = tr2.Update([]byte(key1), []byte(val))
+	_ = tr2.Update([]byte(key3), []byte(val))
+	_ = tr2.Update([]byte(key2), []byte(val))
+	hash2, _ := tr2.Root()
+	fmt.Printf("root hash2: %s\n", base64.StdEncoding.EncodeToString(hash2))
+
+	fmt.Println()
+	tr2.Print(os.Stdout)
+	fmt.Println()
+
+	assert.Equal(t, hash1, hash2)
+}
+
+func TestExtensionNodeToBranchVariant2(t *testing.T) {
```

Comment:
what does variant1 and variant2 mean ? :)

Reply:
changed from variantX to EdgeCaseX as I have found out 2 edge cases